Fix OTA E2E test failures

tgsong opened the pull request:
<!--- Title -->

Description
-----------
<!--- Describe your changes in detail -->
- Directly return from callback if invoked after http backend stopped.
- Never call http connect in callback since the OTA agent thread may stop the data plane.
- Disconnect from S3 when OTA agent cleans up the http data plane.
- Cancel http request if http downloader is already stopped.
- Only free OTA http buffer when OTA agent is stopping.

Checklist:
----------
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Commit:
Fix bugs in OTA HTTP backend

- Directly return from callback if invoked after http backend stopped.
- Never call http connect in callback since the OTA agent thread may
  stop the data plane.
- Disconnect from S3 when OTA agent cleans up the http data plane.

sarenameas commented on libraries/freertos_plus/aws/ota/src/http/aws_iot_ota_http.c:
```diff
@@ -1177,7 +1185,15 @@ OTA_Err_t _AwsIotOTA_CleanupData_HTTP( OTA_AgentContext_t * pAgentCtx )
 
-    memset( &_httpDownloader, 0, sizeof( _httpDownloader_t ) );
+    /* Disconnect from the S3 server. */
+    IotHttpsClient_Disconnect( _httpDownloader.httpConnection.connectionHandle );
```
For possible warnings?
```suggestion
    ( void ) IotHttpsClient_Disconnect( _httpDownloader.httpConnection.connectionHandle );
```

pvyawaha replied:
Can we log the return code here for debug log ?

sarenameas commented on libraries/freertos_plus/aws/ota/src/http/aws_iot_ota_http.c:
```diff
@@ -435,6 +436,12 @@ static void _httpReadReadyCallback( void * pPrivateData,
+    /* Bail out if this callback is invoked after http downloader stopped. */
+    if( _httpDownloader.state == OTA_HTTP_STOPPED )
+    {
+        return;
```
```suggestion
        IotHttpsClient_CancelResponseAsync( responseHandle );
        return;
```

sarenameas replied:
But this will invoke the errorcallback saying saying the user cancelled.

sarenameas replied:
Per offline conversation `IotHttpsClient_CancelRequestAsync( requestHandle );` will also be added to _httpAppendHeaderCallback().

Commit:
More bug fixes in OTA http backend

- Cancel http request if http downloader is already stopped
- Only free OTA http buffer when OTA agent is stopping

Commit:
Address PR feedback

Commit:
Change to debug log for http disconnect in OTA
